feat: Added SecretTag feature with "eye" icon toggle

umutdz opened the pull request:
# Description

Added a new `SecretTag` feature that provides secure input handling for sensitive information like passwords and API keys. This implementation includes:

## Features
- Masked input field that displays "•" characters instead of actual text
- Optional toggle button with an eye icon (👁) to show/hide the sensitive content
- Safe string representation to prevent accidental exposure in logs
- Support for both simple form and dataclass usage

## Implementation Details
- Added `SecretTag` class extending the base `Tag`
- Created `SecretEntryFrame` widget for Tkinter UI
- Added support in `widgetize` method for both TUI and GUI
- Implemented hashable support for use with Python's `Annotated`

## Example Usage
```python
# Simple form
m.form({
    "Password": SecretTag(""),
    "API Key": SecretTag("", show_toggle=True)  # with eye icon
})

# With dataclass
@dataclass
class Credentials:
    password: Annotated[str, SecretTag()]
    api_key: Annotated[str, SecretTag(show_toggle=True)]
```

## Tests
Added unit tests covering:
- Secret masking functionality
- Toggle visibility feature
- Safe string representation
- Empty secret handling
- Annotation default behavior

e3rd commented:
Your contribution seems amazing! Had I not implement it today too, along with some other features, I would accept it immediately.
Would you mind if I took only parts of your code and integrate them? (IE. the dunder methods are great.)

A test does not pass for me, GUI outputs the original value instead of the password, but TUI is much better than mine.

```
# Simple form
print(m.form({
    "Password": SecretTag(""),
    "API Key": SecretTag("", show_toggle=True)  # with eye icon
})) # {'Password': '', 'API Key': ''} -> the original value
```

umutdz commented:
Happy to hear that! I can update it according to your implementation, could you please give me more info about your request?

e3rd commented:
Thanks! I've pushed the new commit to the dev brach, picked few features and put you as a co-author (okay like that ?). Do you like the code functionality now? (A default is reversed etc.) If you'd like to elaborate, wanna implement the shortcut to the GUI as well or something?

(For me, textual shortcut ctrl+h does not work for me as it works like backspace in my bash but that might be a bash or textual library problem.)

umutdz commented:
Yes, I’m happy to be added as a co-author. The code functionality looks great, and I’ll take a look at what we can do for the GUI shortcut.

umutdz commented:
@e3rd I fixed the shortcut and added TkInterface as well, check it out pls

"""
        Keyboard Shortcuts:
        - Ctrl+S: Save form
        - Ctrl+Q: Quit
        - Ctrl+H: Show this help
        - Ctrl+T: Toggle visibility of password fields
        - Enter: Submit form
        - Escape: Cancel
        """

e3rd commented:
Wow, you're quick! I fear I'd like you to elaborate the dev branch, since there are your changes, not here.
Rather than building a help, that may include some options that are not present (ex. the form not having Password field), put the Ctrl+T hint to the toggle-tooltip button only. I think no need to adding Ctrl+S for sumbitting, as Ctrl+Enter already does that and I'd like to leave Ctrl+S for the user (future usage).
You're great and no need to hurry with the changes!

e3rd commented:
(I close this for now as we should continue now on the dev branch.)

umutdz commented:
So, should I open the PR to the dev branch for the shortcut feature? (without unnecessary like ctrl+s)

e3rd commented:
Yes please :)

On Fri, Mar 14, 2025, 23:34 umut deniz ***@***.***> wrote:

> So, should I open the PR to the dev branch for the shortcut feature?
> (without unnecessary like ctrl+s)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CZ-NIC/mininterface/pull/14#issuecomment-2725912681>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABU64F5HIVYHTLU3J2VIUDL2UNKOHAVCNFSM6AAAAABZA4TNCWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRVHEYTENRYGE>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
> [image: umutdz]*umutdz* left a comment (CZ-NIC/mininterface#14)
> <https://github.com/CZ-NIC/mininterface/pull/14#issuecomment-2725912681>
>
> So, should I open the PR to the dev branch for the shortcut feature?
> (without unnecessary like ctrl+s)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CZ-NIC/mininterface/pull/14#issuecomment-2725912681>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABU64F5HIVYHTLU3J2VIUDL2UNKOHAVCNFSM6AAAAABZA4TNCWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRVHEYTENRYGE>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>
